integrated_channels/canvas/client.py

```diff
@@ -62,36 +64,110 @@ def create_content_metadata(self, serialized_data):
+        course_id = self._get_course_id_from_integration_id(integration_id)
+
+        url = '{}/api/v1/courses/{}'.format(
+            self.enterprise_configuration.canvas_base_url,
+            course_id,
+        )
+        put_response = self.session.put(url, data=data)
+        put_response.raise_for_status()
+
+        return put_response.status_code, put_response.text
+
+    def _delete(self, data):
+        """
+        Make a DELETE request using the session object to the Canvas course delete endpoint.
+
+        Args:
+            data (bytearray): The json encoded payload to DELETE. This also contains the integration
+            ID used to match a course with a course ID.
+        """
+        self._create_session()
+
+        integration_id = self._get_integration_id_from_transmition_data(data)
+        course_id = self._get_course_id_from_integration_id(integration_id)
+
+        url = '{}/api/v1/courses/{}'.format(
+            self.enterprise_configuration.canvas_base_url,
+            course_id,
+        )
+        delete_response = self.session.delete(url, data='{"event":"delete"}')
+        delete_response.raise_for_status()
+
+        return delete_response.status_code, delete_response.text
+
+    def _get_integration_id_from_transmition_data(self, data):
+        """
+        Retrieve the integration ID string from the encoded transmission data and apply appropriate
+        error handling.
+
+        Args:
+            data (bytearray): The json encoded payload intended for a Canvas endpoint.
+        """
+        if not data:
+            raise CanvasClientError("No data to transmit.")
+        try:
+            integration_id = json.loads(
+                data.decode("utf-8")
+            )['course']['integration_id']
+        except KeyError:
+            raise CanvasClientError("Could not transmit data, no integration ID present.")
+        except AttributeError:
+            raise CanvasClientError("Unable to decode data.")
+
+        return integration_id
+
+    def _get_course_id_from_integration_id(self, integration_id):
+        """
+        To obtain course ID we have to request all courses associated with the integrated
+        account and match the one with our integration ID.
+
+        Args:
+            integration_id (string): The ID retrieved from the transmission payload.
+        """
+        url = "{}/api/v1/accounts/{}/courses/".format(
+            self.enterprise_configuration.canvas_base_url,
+            self.enterprise_configuration.canvas_account_id
+        )
+        all_courses_response = self.session.get(url).json()
+        course_id = None
+        for course in all_courses_response:
+            if course['integration_id'] == integration_id:
+                course_id = course['id']
+                break
+
+        if not course_id:
+            raise CanvasClientError("No Canvas courses found with associated integration ID.")
```

Comment:
nit: ah also to those reading these logs (or newrelic etc) it will be awesome to have the integration_id in the logs

Reply:
Absolutely!